darkcluster/src/main/java/com/linkedin/darkcluster/util/GuaranteedRateLimiter.java

```diff
@@ -0,0 +1,279 @@
+/*
+   Copyright (c) 2021 LinkedIn Corp.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.darkcluster.util;
+
+import com.linkedin.common.callback.Callback;
+import com.linkedin.common.util.None;
+import com.linkedin.r2.transport.http.client.AsyncRateLimiter;
+import com.linkedin.r2.transport.http.client.ratelimiter.Rate;
+import com.linkedin.util.ArgumentUtil;
+import com.linkedin.util.clock.Clock;
+import java.util.NoSuchElementException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nonnull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * A simple {@link AsyncRateLimiter} implementation that re-issues permits at every specified period of time.
+ * A submitted callback's #onError is invoked with {@link RejectedExecutionException} if the current buffered
+ * callbacks exceeds to the maximum allowed by the implementation. Based heavily on SmoothRateLimiter by Sean Sheng.
+ *
+ * @author Lester Haynes
+ * @author Sean Sheng
+ * @author Francesco Capponi (fcapponi@linkedin.com)
+ */
+public class GuaranteedRateLimiter implements AsyncRateLimiter
```

Comment:
Would it be possible to extend the SmoothRateLimiter to avoid a lot of code duplication? Or maybe pull out the EventLoop class from the inner class, so both SmoothRateLimiter and GuaranteedRateLimiter can use it.

Reply:
Agreed, will want to deduplicate a lot of this once the overall design is approved